registry/registry.go

```diff
@@ -168,42 +156,54 @@ func (r *RouteRegistry) Unregister(uri route.Uri, endpoint *route.Endpoint) {
 		return
 	}
 
-	r.unregister(uri, endpoint)
+	routeKey := uri.RouteKey()
+	endpointUnregisteredResult, pool := r.unregisterEndpoint(routeKey, endpoint)
+	if pool == nil {
+		return
+	}
+
+	r.reporter.CaptureUnregistryMessage(endpoint, string(endpointUnregisteredResult))
+	if r.logger.Enabled(context.Background(), slog.LevelInfo) {
+		r.logger.Info(string(endpointUnregisteredResult), buildSlogAttrs(routeKey, endpoint)...)
+	}
 
-	r.reporter.CaptureUnregistryMessage(endpoint)
+	routeUnregisteredResult := r.deleteRouteWithoutEndpoint(routeKey, pool)
+	switch routeUnregisteredResult {
+	case route.RouteUnregistered:
+		r.reporter.CaptureUnregistryMessage(endpoint, string(routeUnregisteredResult))
+		if r.logger.Enabled(context.Background(), slog.LevelInfo) {
+			r.logger.Info(string(routeUnregisteredResult), slog.Any("uri", routeKey))
+		}
+	}
 }
 
-func (r *RouteRegistry) unregister(uri route.Uri, endpoint *route.Endpoint) {
+func (r *RouteRegistry) unregisterEndpoint(routeKey route.Uri, endpoint *route.Endpoint) (route.PoolRemoveEndpointResult, *route.EndpointPool) {
 	r.Lock()
 	defer r.Unlock()
 
-	uri = uri.RouteKey()
+	pool := r.byURI.Find(routeKey)
+	if pool == nil {
+		return route.EndpointNotUnregistered, nil
+	}
+	return pool.Remove(endpoint), pool
+}
 
-	pool := r.byURI.Find(uri)
-	if pool != nil {
-		endpointRemoved := pool.Remove(endpoint)
-		if endpointRemoved {
-			if r.logger.Enabled(context.Background(), slog.LevelInfo) {
-				r.logger.Info("endpoint-unregistered", buildSlogAttrs(uri, endpoint)...)
-			}
-		} else {
-			if r.logger.Enabled(context.Background(), slog.LevelInfo) {
-				r.logger.Info("endpoint-not-unregistered", buildSlogAttrs(uri, endpoint)...)
-			}
-		}
+func (r *RouteRegistry) deleteRouteWithoutEndpoint(routeKey route.Uri, pool *route.EndpointPool) route.PoolRemoveRouteResult {
+	r.Lock()
+	defer r.Unlock()
 
-		if pool.IsEmpty() {
-			if r.EmptyPoolResponseCode503 && r.EmptyPoolTimeout > 0 {
-				if time.Since(pool.LastUpdated()) > r.EmptyPoolTimeout {
-					r.byURI.Delete(uri)
-					r.logger.Info("route-unregistered", slog.Any("uri", uri))
-				}
-			} else {
-				r.byURI.Delete(uri)
-				r.logger.Info("route-unregistered", slog.Any("uri", uri))
+	if pool.IsEmpty() {
+		if r.EmptyPoolResponseCode503 && r.EmptyPoolTimeout > 0 {
+			if time.Since(pool.LastUpdated()) > r.EmptyPoolTimeout {
+				r.byURI.Delete(routeKey)
+				return route.RouteUnregistered
 			}
+		} else {
+			r.byURI.Delete(routeKey)
+			return route.RouteUnregistered
```

Comment:
```suggestion
		if !(r.EmptyPoolResponseCode503 && r.EmptyPoolTimeout > 0) 
		    || (r.EmptyPoolResponseCode503 && r.EmptyPoolTimeout > 0 
		        && if time.Since(pool.LastUpdated()) > r.EmptyPoolTimeout) {
				r.byURI.Delete(routeKey)
				return route.RouteUnregistered
			}
```

Reply:
Applied